### ISSUE TITLE
Inconsistent error handling when micronaut.server.max-request-size exceeded

### ISSUE THREAD

**Issue by Sandmania** [micronaut-projects/micronaut-core#4864]
### Task List

- [x] Steps to reproduce provided
- [ ] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

## Description

There seems to be some inconsistency in handling `ContentLengthExceededException` when posting files with multipart forms. The variations I've observed in behaviour depend on:
* How much `micronaut.server.max-request-size` was exceeded.
And
* Is an error handler used and which status code that error handler returns.

Also the client being used seems to play a role, which makes me question if this a bug in mn or have I just done something dumb.

Below you'll find three test cases for uploading files and handling errors:
* No error handler (So what ever mn does by default)
* Local error handler returning HTTP 413
* Local error handler returning HTTP 500

Each case is tested with three different clients: Firefox, HTTPie and curl. With curl I can consistently see expected behaviour, but with Firefox and HTTPie I can consistently see the actual behaviour.    
Unfortunately I was not able to write JUnit tests with which I could replicate these.

## Steps to Reproduce

Sample application can be found [from this repository](https://github.com/Sandmania/micronaut-request-size-exceeded).

### Case: No custom error handler

1. Start sample application by running: `./mvnw mn:run`
2. See client specific steps below

| Firefox                                                                | HTTPie                                                                                     | curl                                                                          |
|------------------------------------------------------------------------|--------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------|
| Navigate to http://localhost:8080/                                     | `http --print=hHb --form http://localhost:8080/upload files@src/test/resources/900k.dummy` | `curl http://localhost:8080/upload -F 'files=@src/test/resources/900k.dummy'` |
| Under **None or global** select `900k.dummy` from `src/test/resources` |                                                                                            |                                                                               |
| Push `Upload attachments`                                              |                                                                                            |                                                                               |

#### Expected behaviour

Regardles of client used, all return `HTTP 413` error with message `The content length [921600] exceeds the maximum allowed content length [614400]`.    

#### Actual behaviour

| Firefox (NOK)                                    | HTTPie (NOK)                                     | curl (OK)                                                                                                 |
|--------------------------------------------------|--------------------------------------------------|-----------------------------------------------------------------------------------------------------------|
| POST to  `/upload`  endpoint hangs indefinetely. | POST to  `/upload`  endpoint hangs indefinetely. | `HTTP 413` with message `The content length [921803] exceeds the maximum allowed content length [614400]` |

It's worth noting, that if you choose a slightly smaller file (say, the 800k.dummy provided) you can observe the expected behaviour with all clients.

### Case: Local error handler that returns HTTP 413

The expected behaviour and actual behaviour are consistent with previous case. Only the endpoint called differes as it has a local error handler.

1. Start sample application by running: `./mvnw mn:run`
2. See client specific steps below

| Firefox                                                                           | HTTPie                                                                                                        | curl                                                                                                |
|-----------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------|
| Navigate to http://localhost:8080/                                                | `http --print=hHb --form http://localhost:8080/uploadLocalErrorHandling files2@src/test/resources/900k.dummy` | `curl -v http://localhost:8080/uploadLocalErrorHandling -F 'files2=@src/test/resources/900k.dummy'` |
| Under **Local with HttpStatus 413** select `900k.dummy` from `src/test/resources` |                                                                                                               |                                                                                                     |
| Push `Upload attachments`                                                         |                                                                                                               |                                                                                                     |                                                   |                                                                                                              |                                                                                                 |                                          |                                                                                            |                                                                               |

#### Expected behaviour

Regardles of client used, all return `HTTP 413` error with message `The content length [921600] exceeds the maximum allowed content length [614400]`.    

#### Actual behaviour

| Firefox (NOK)                                                      | HTTPie (NOK)                                                       | curl (OK)                                                                                                 |
|--------------------------------------------------------------------|--------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------|
| POST to  `/uploadLocalErrorHandling`  endpoint hangs indefinetely. | POST to  `/uploadLocalErrorHandling`  endpoint hangs indefinetely. | `HTTP 413` with message `The content length [921803] exceeds the maximum allowed content length [614400]` |

It's worth noting, that if you choose a slightly smaller file (say, the 800k.dummy provided) you can observe the expected behaviour with all clients.

### Case: Local error handler that returns HTTP 500

1. Start sample application by running: `./mvnw mn:run`
2. See client specific steps below

| Firefox                                                                           | HTTPie                                                                                                                         | curl                                                                                                                 |
|-----------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------|
| Navigate to http://localhost:8080/                                                | `http --print=hHb --form http://localhost:8080/uploadLocalErrorHandlingReturnServerError files3@src/test/resources/900k.dummy` | `curl -v http://localhost:8080/uploadLocalErrorHandlingReturnServerError -F 'files3=@src/test/resources/900k.dummy'` |
| Under **Local with HttpStatus 500** select `900k.dummy` from `src/test/resources` |                                                                                                                                |                                                                                                                      |
| Push `Upload attachments`                                                         |                                                                                                                                |                                                                                                                      |

#### Expected behaviour

Regardles of client used, all return `HTTP 500` error with message `The content length [921600] exceeds the maximum allowed content length [614400]`.    

#### Actual behaviour

| Firefox (NOK)              | HTTPie (NOK)                                                                   | curl (OK)                                                                                                 |
|----------------------------|--------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------|
| `The connection was reset` | `ConnectionError: ('Connection aborted.', BrokenPipeError(32, 'Broken pipe'))` | `HTTP 500` with message `The content length [921803] exceeds the maximum allowed content length [614400]` |

### Environment Information

- **Operating System**: macOs Big Sur 11.
- **Micronaut Version:** 2.2.0
- **JDK Version:** 1.8.0_232, vendor: AdoptOpenJDK
- **Browser:** Firefox 85.0 
- **curl:** 7.64.1 
- **HTTPie:** 2.3.0

### Example Application

Sample application can be found [from this repository](https://github.com/Sandmania/micronaut-request-size-exceeded).

**Comment by Sandmania** [comment]
I see the same kind of behavior around `micronaut.server.multipart.max-file-size`.

**Comment by ahimik** [comment]
Same here.

**Comment by morki** [comment]
We have the same problem.

**Comment by morki** [comment]
The 413 error is displayed in network tab in chrome dev tools after stopping of Micronaut application. Until then, it hangs indefinitely.

**Comment by morki** [comment]
Configuration:
```yaml
micronaut:
  server:
    max-request-size: 1GB
    multipart:
      max-file-size: XXXMB
      mixed: true
      disk: true
      location: /tmp
```

Uploading the same file with size `164.7MB` and changing **only** `micronaut.server.multipart.max-file-size` will result in:

|`max-file-size`|Result|Expected behaviour|
|---|---|---|
|200M|File is successfully uploaded|yes|
|150M|Server returns `413 Entity Too Large`|yes|
|100M|Request indefinitely hangs|no|

**Comment by morki** [comment]
There is really something broken with multipart handling. With the same configuration as above and arbitrary `max-file-size`, handling of any POST requests with `Content-Type: application/x-www-form-urlencoded` begin to fail, for example `LoginController` from `micronaut-security` project:  

```
###
POST http://127.0.0.1:8080/login
Content-Type: application/x-www-form-urlencoded

username=X&password=Y
```

Response:
```
{
  "message": "Internal Server Error",
  "logref": null,
  "path": null,
  "_links": {
    "self": {
      "href": "/login",
      "templated": false,
      "profile": null,
      "deprecation": null,
      "title": null,
      "hreflang": null,
      "type": null,
      "name": null
    }
  },
  "_embedded": {
    "errors": [
      {
        "message": "Internal Server Error: refCnt: 0, decrement: 1",
        "logref": null,
        "path": null,
        "_links": {},
        "_embedded": {}
      }
    ]
  }
}
```

After removing `disk: true` and `mixed: true` it works normally.

**Comment by morki** [comment]
@yawkat @graemerocher this is now stopping us from production, is there anything we can do to move it forward?

**Comment by yawkat** [comment]
ok i will take a look, since it is blocking you.

**Comment by yawkat** [comment]
This issue report uses an old micronaut version and I can't fully reproduce the hang. I've found a hang that looks similar though, and fixed that in #6696. 

However @morki I'm not confident this will fix your specific issue. I think your issue looks similar but may be different. If you can still reproduce this with the patch, please create a new issue with a test case.

**Comment by morki** [comment]
Hi @yawkat, thank you very much for taking time for this issue. I prepared minimal reproduction in this repository: https://github.com/morki/micronaut-multipart-issue 

There are 2 described issues:

1) On route "/" there is login form for micronaut security login endpoint. You can write anything in those fields and click submit. After that, error `Internal Server Error: refCnt: 0, decrement: 1` is thrown. But when you remove configuration `mixed: true` and `disk: true`, everything is running fine.

2) On route "/upload" there is multipart form with file selection. Please select file `big.txt` contained in the root of this repository and click submit. Request in browser will hang. But, when you stop the server, the `Entity too large` response is sent.

Again thank you very much for your time :)

**Comment by yawkat** [comment]
@morki yes the first is definitely a different issue, I will check the second as well. Please submit this as a new bug (a single bug for both is enough) for tracking purposes, as this issue will be closed by #6696.

**Comment by morki** [comment]
Done in https://github.com/micronaut-projects/micronaut-core/issues/6705
